Components/IconTypes/IconType_customvalue/customvalue.lua

```diff
@@ -0,0 +1,130 @@
+﻿local TMW = TMW
+if not TMW then return end
+
+local L = TMW.L
+
+local print = TMW.print
+
+
+local Type = TMW.Classes.IconType:New("customvalue")
+
+Type.name = L["ICONMENU_CUSTOMVALUE"]
+Type.desc = L["ICONMENU_CUSTOMVALUE_DESC"]
+Type.menuIcon = "Interface/Icons/inv_misc_punchcards_white"
+
+Type.hasNoGCD = true
+Type.barIsTimer = false
+
+local STATE_SUCCEED = TMW.CONST.STATE.DEFAULT_SHOW
+local STATE_FAIL = TMW.CONST.STATE.DEFAULT_HIDE
+
+Type:SetAllowanceForView("icon", false)
+Type:SetModuleAllowance("IconModule_TimerBar_Overlay", false)
+Type:SetModuleAllowance("IconModule_CooldownSweep", false)
+Type:UsesAttributes("texture")
+
+Type:UsesAttributes("value")
+Type:UsesAttributes("maxValue")
+Type:UsesAttributes("valueColor")
+Type:UsesAttributes("state")
+
+local BarColors = {{r=0, g=0, b=1, a=1}, {r=0, g=1, b=1, a=1}, {r=0, g=1, b=0, a=1}}
+
+Type:RegisterIconDefaults{
+	-- Lua code to be evaluated
+	LuaCode			="",
+
+	-- Currently displayed value
+	value			=0,
+	
+	-- Maximum value displayed
+	maxValue		=0,
+	
+	-- Bar color scheme
+	valueColor		=BarColors,
+	
+	-- Initial state of icon (set as fail since we have no code yet)
+	state			=STATE_FAIL,
+
+}
+
+
+Type:RegisterConfigPanel_XMLTemplate(165, "TellMeWhen_CustomValue", {
+	implementConditions = true,
+})
+
+Type:RegisterConfigPanel_XMLTemplate(100, "TellMeWhen_IconStates", {
+	[STATE_SUCCEED] = { text = "|cFF00FF00" .. L["ICONMENU_CUSTOMVALUE_OK"], },
+	[STATE_FAIL] =    { text = "|cFFFF0000" .. L["ICONMENU_CUSTOMVALUE_ERROR"], },
+})
+
+
+local function CustomValue_OnUpdate(icon, time)    
+
+	local value, maxVaule, func = icon.value, icon.maxValue, icon.luaFunc
+	
+
+	if func == nil then
+		icon:SetInfo("state", STATE_FAIL)
+		return
+	end
+	value, maxValue = func()
+	value = tonumber(value)
+	maxValue = tonumber(maxValue)
+	if value == nil or maxValue == nil then
+		icon:SetInfo("state", STATE_FAIL)
+		return
+	end
+	if value < 0 then
+		value = 0
+	end
+	if maxValue < 0 then
+		icon:SetInfo("state", STATE_FAIL)
+		return
+	end
+	if value > maxValue then
+		maxValue = value
+	end	
```

Comment:
Traditionally one would constrain the `value` to be within its limits (`maxValue`), not the other way around.

Reply:
If the value was constrained, there would be no indication that the displayed value has been capped. By changing the limit, the displayed value would still match the calculated value. Of course, none of this matters if the max value is properly set in the first place.